core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -3135,12 +3135,40 @@ private void setAclFromEntry(File.SetAclEntry entry) throws FileDoesNotExistExce
         .lockFullInodePath(entry.getId(), InodeTree.LockMode.WRITE)) {
       setAclInternal(RpcContext.NOOP, SetAclAction.fromProto(entry.getAction()), inodePath,
           entry.getEntriesList().stream().map(AclEntry::fromProto).collect(Collectors.toList()),
-          entry.getOpTimeMs(), options);
+          true, entry.getOpTimeMs(), options);
+    }
+  }
+
+  private void setUfsAcl(RpcContext rpcContext, LockedInodePath inodePath)
+      throws InvalidPathException, AccessControlException {
+    Inode<?> inode = inodePath.getInodeOrNull();
+
+    checkUfsMode(inodePath.getUri(), OperationType.WRITE);
+    MountTable.Resolution resolution = mMountTable.resolve(inodePath.getUri());
+    String ufsUri = resolution.getUri().toString();
+    try (CloseableResource<UnderFileSystem> ufsResource = resolution.acquireUfsResource()) {
+      UnderFileSystem ufs = ufsResource.get();
+      if (ufs.isObjectStorage()) {
+        LOG.warn("SetACL is not supported to object storage UFS via Alluxio. "
+            + "UFS: " + ufsUri + ". This has no effect on the underlying object.");
+      } else {
+        try {
+          ufs.setAcl(ufsUri, inode.getACL());
+          if (inode.isDirectory()) {
+            InodeDirectory inodeDirectory = (InodeDirectory) inode;
+            ufs.setAcl(ufsUri, inodeDirectory.getDefaultACL());
+          }
+          // TODO(david): set ufs's default ACL as well.
+        } catch (IOException e) {
+          throw new AccessControlException("Could not setAcl for UFS file" + ufsUri + ".");
```

Comment:
This doesn't distinguish between a ufs which fails to set it and a ufs which does not support acls, right?

Reply:
ufs does not support acls now converted to noop, so all exceptions thrown are for ufs which fails to set it.